x/meshsecurity/keeper/stake.go

```diff
@@ -2,10 +2,66 @@ package keeper
 
 import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/errors"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+
+	types "github.com/osmosis-labs/mesh-security-sdk/x/meshsecurity/types"
 )
 
-func (k Keeper) Delegate(pCtx sdk.Context, actor sdk.AccAddress, valAddr sdk.ValAddress, amt sdk.Coin) error {
-	panic("not implemented, yet")
+// Delegate delegate mints new "virtual" bonding tokens and delegates them to the given validator.
+// The amount minted is added to the SupplyOffset, when supported.
+func (k Keeper) Delegate(pCtx sdk.Context, actor sdk.AccAddress, valAddr sdk.ValAddress, amt sdk.Coin) (sdk.Dec, error) {
+	if amt.Amount.IsZero() || amt.Amount.IsNegative() {
+		return sdk.ZeroDec(), errors.ErrInvalidRequest.Wrap("amount")
+	}
+
+	// Ensure staking constraints
+	bondDenom := k.staking.BondDenom(pCtx)
+	if amt.Denom != bondDenom {
+		return sdk.ZeroDec(), errors.ErrInvalidRequest.Wrapf("invalid coin denomination: got %s, expected %s", amt.Denom, bondDenom)
+	}
+	validator, found := k.staking.GetValidator(pCtx, valAddr)
+	if !found {
+		return sdk.ZeroDec(), stakingtypes.ErrNoValidatorFound
+	}
+
+	// Ensure MS constraints:
+	newTotalDelegatedAmount := k.getTotalDelegatedAmount(pCtx, actor).Add(amt.Amount)
+	max := k.GetMaxCapLimit(pCtx, actor).Amount
+	if newTotalDelegatedAmount.GT(max) {
+		return sdk.ZeroDec(), types.ErrMaxCapExceeded
```

Comment:
Do you want a different error when maxcap is 0.

One is "You could do 1000, but tried 1200". The other is "You haven't been authorized at all".
Different error messages, or including the cap amount in the error message would help debugging

Reply:
👍 Adding numbers here will be helpful. Good idea!
Important but authorization should be handled before entering this method as separation of concerns. I will add a note to the doc.